strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+        sq = symp.expand(symp.squeezing(r, phi), k, self.nlen)
+        self.means = update_means(self.means, sq, self.from_xp)
+        self.covs = update_covs(self.covs, sq, self.from_xp)
+
+    def mbsqueeze(self, k, r, phi, r_anc, eta_anc, avg):
+        r"""Squeeze mode ``k`` by the amount ``r*exp(1j*phi)`` using measurement-based squeezing.
+
+        Either the average map, described by a Gaussian CPTP transformation, or a single-shot map
+        with ancillary measurement outcomes can be simulated.
+
+        Args:
+            k (int): mode to be squeezed
+            r (float): target squeezing magnitude
+            phi (float): target squeezing phase
+            r_anc (float): squeezing magnitude of the ancillary mode
+            eta_anc(float): detection efficiency of the ancillary mode
+            avg (bool): whether to apply the average map or single-shot
+        Returns:
+            float: if single-shot map selected, returns the measurement outcome of the ancilla
+        """
+
+        if self.active[k] is None:
+            raise ValueError("Cannot squeeze mode, mode does not exist")
+
+        # antisqueezing corresponds to an extra phase shift
+        if r < 0:
+            phi += np.pi
+        r = np.abs(r)
+        # beamsplitter angle
+        theta = np.arccos(np.exp(-r))
+        self.phase_shift(-phi / 2, k)
+
+        # Construct (X,Y) for Gaussian CPTP of average map
+        if avg:
+            X = np.diag([np.cos(theta), 1 / np.cos(theta)])
+            Y = np.diag(
+                [
+                    (np.sin(theta) ** 2) * (np.exp(-2 * r_anc)),
+                    (np.tan(theta) ** 2) * (1 - eta_anc) / eta_anc,
+                ]
+            )
+            Y *= self.hbar / 2
+            X2, Y2 = self.expandXY([k], X, Y)
+            self.apply_channel(X2, Y2)
+
+        # Add new ancilla mode, interfere it and measure it
+        # Delete ancilla mode from active list
+        elif not avg:
+            self.add_mode()
+            new_mode = self.nlen - 1
+            self.squeeze(r_anc, 0, new_mode)
+            self.beamsplitter(theta, 0, k, new_mode)
+            self.loss(eta_anc, new_mode)
+            self.phase_shift(np.pi / 2, new_mode)
+            val = self.homodyne(new_mode)
+            self.del_mode(new_mode)
+            self.covs = np.delete(self.covs, [2 * new_mode, 2 * new_mode + 1], axis=1)
+            self.covs = np.delete(self.covs, [2 * new_mode, 2 * new_mode + 1], axis=2)
+            self.means = np.delete(self.means, [2 * new_mode, 2 * new_mode + 1], axis=1)
+            self.nlen = self.nlen - 1
+            self.from_xp = from_xp(self.nlen)
+            self.to_xp = to_xp(self.nlen)
+            self.active = self.active[:new_mode]
+            prefac = -np.tan(theta) / np.sqrt(2 * self.hbar * eta_anc)
+            self.displace(prefac * val[0][0], np.pi / 2, k)
+
+        self.phase_shift(phi / 2, k)
+
+        if not avg:
+            return val
+
+        return None
```

Comment:
Maybe this method could be split into two based on the `avg` boolean (separate average map and single-shot cases).

Although there would be some code duplication, it is somewhat unexpected that based on the value of `avg` we either return the measurement result or simply just update the circuit.

Reply:
As you will see in a future PR, at the frontend level, I just have an sf.ops called MbSqueeze, and it has the same bool option for avg. If I break up the function here, should I also break the front end ops method?

I know it is bizarre to have the method return different types, but it's motivated by the physics in that this is a gate that requires an ancillary resource state, so you can either write down what the gate does on average (averaging over the measurement outcomes of the ancilla), or write down what the state does in a single-shot instance.

Reply:
Hmm, this might be more connected to what the preferred UI could be for users. If it's more likely, that people would like to perform measurement-based squeezing and this is a small detail only regarding the operation, then maybe it's good to be kept as is in the frontend. Otherwise, it could also be made into two separate operations.

Under the hood, separation would just help with future maintenance of the code.

Usually from a code quality point of view, when there's a boolean argument which greatly influences the logic of the function, it tells that the function is doing at least two or more things (which sort of goes against the [Single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) extended to functions/methods).

Reply:
Thanks for the info! I will separate it in the backend then, but leave it as one call in the frontend.

Reply:
Done.